### ISSUE TITLE
How to start docker service only after LVM is up?

### ISSUE THREAD

**Issue by ueriku** [moby/moby#34343]
**Description**

After vm reboot, docker(direct-lvm mode) service fails to start occasionally.
How to start docker service only after LVM is up?

```
# systemctl status docker
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since 月 2017-07-31 16:20:44 JST; 17h ago
     Docs: https://docs.docker.com
  Process: 872 ExecStart=/usr/bin/dockerd --storage-driver=devicemapper --storage-opt=dm.thinpooldev=/dev/mapper/docker-thinpool --storage-opt dm.use_deferred_removal=true --experimental (code=exited, status=1/FAILURE)
 Main PID: 872 (code=exited, status=1/FAILURE)
   Memory: 31.6M
   CGroup: /system.slice/docker.service

 7月 31 16:20:40 aqp-mgr001 systemd[1]: Starting Docker Application Container Engine...
 7月 31 16:20:44 aqp-mgr001 dockerd[872]: time="2017-07-31T16:20:44+09:00" level=warning msg="Running experimental build"
 7月 31 16:20:44 aqp-mgr001 dockerd[872]: time="2017-07-31T16:20:44.510169372+09:00" level=info msg="libcontainerd: new containerd process, pid: 1094"
 7月 31 16:20:44 aqp-mgr001 dockerd[872]: Error starting daemon: error initializing graphdriver: devicemapper: Non existing device docker-thinpool
 7月 31 16:20:44 aqp-mgr001 systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
 7月 31 16:20:44 aqp-mgr001 systemd[1]: Failed to start Docker Application Container Engine.
 7月 31 16:20:44 aqp-mgr001 systemd[1]: Unit docker.service entered failed state.
 7月 31 16:20:44 aqp-mgr001 systemd[1]: docker.service failed.
```

**Steps to reproduce the issue:**
1. create lvm
```
pvcreate /dev/sdc
vgcreate docker /dev/sdc
lvcreate --wipesignatures y -n thinpool docker -l 95%VG
lvcreate --wipesignatures y -n thinpoolmeta docker -l 1%VG
lvconvert -y --zero n -c 512K --thinpool docker/thinpool --poolmetadata docker/thinpoolmeta
tee /etc/lvm/profile/docker-thinpool.profile <<-'EOF'
activation {
    thin_pool_autoextend_threshold=80
    thin_pool_autoextend_percent=20
}
EOF
lvchange --metadataprofile docker-thinpool docker/thinpool
lvs -o+seg_monitor
```
2. edit unit file(/usr/lib/systemd/system/docker.service)
```
[Unit]
Description=Docker Application Container Engine
Documentation=https://docs.docker.com
After=network.target firewalld.service lvm2-monitor.service

[Service]
Type=notify
# the default is not to use systemd for cgroups because the delegate issues still
# exists and systemd currently does not support the cgroup feature set required
# for containers run by docker
ExecStart=/usr/bin/dockerd  --storage-driver=devicemapper --storage-opt=dm.thinpooldev=/dev/mapper/docker-thinpool --storage-opt dm.use_deferred_removal=true --experimental
ExecReload=/bin/kill -s HUP $MAINPID
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=infinity
LimitNPROC=infinity
LimitCORE=infinity
# Uncomment TasksMax if your systemd version supports it.
# Only systemd 226 and above support this version.
#TasksMax=infinity
TimeoutStartSec=0
# set delegate yes so that systemd does not reset the cgroups of docker containers
Delegate=yes
# kill only the docker process, not all processes in the cgroup
KillMode=process

[Install]
WantedBy=multi-user.target
```
3. enable & start docker service
```
systemctl daemon-reload
systemctl start docker
systemctl enable docker
```
4. reboot VM

**Additional information you deem important (e.g. issue happens only occasionally):**

issue happens only occasionally(once every 5-10 times)

**Output of `docker version`:**

```
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:05:44 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:05:44 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 7
 Running: 7
 Paused: 0
 Stopped: 0
Images: 39
Server Version: 17.03.1-ce
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 12.78 GB
 Data Space Total: 130.6 GB
 Data Space Available: 117.8 GB
 Metadata Space Used: 12.39 MB
 Metadata Space Total: 1.372 GB
 Metadata Space Available: 1.359 GB
 Thin Pool Minimum Free Space: 13.06 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: active
 NodeID: ne2mqkj4659s6bhjc2jhk7ok9
 Is Manager: true
 ClusterID: kc5qzj456r563xqx2f0nc81gz
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.0.0.4
 Manager Addresses:
  10.0.0.4:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.21.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 13.7 GiB
Name: mgr1
ID: NY7V:6ZGH:N22Q:3BXX:6XC3:RNUT:KHNH:SK6B:IWIU:XM7H:O5LI:23AO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy:
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Azure

**Comment by wrfly** [comment]
You need to re-active the device by `sudo lvchange -ay docker/thinpool` after a reboot.

**Comment by fsignoret** [comment]
Hello wrfly,
I have the same issue as described above, and your command does fix it.
However, how to make a fix for good, without the need of manual intervention? (the machine is on the Amazon cloud and is shut down at least everyday).
Thank you!
Regards,
Florian

**Comment by wrfly** [comment]
@fsignoret You can write that command to `/etc/rc.local` for a try :)

**Comment by vikramkhatri** [comment]
Even though this issue was raised in July 2017, this is still relevant as I faced same issue where I was getting error:
```
dockerd: Error starting daemon: error initializing graphdriver: devicemapper: Non existing device docker-thinpool
```
My Docker version is 18.03 and this issue seems to have been fixed long time back but it still shows up occasionally. After going through many posts, I made slight modifications in my `/lib/systemd/system/docker.service` and now it works perfect for me. 

```
# cat /lib/systemd/system/docker.service 
[Unit]
Description=Docker Application Container Engine
Documentation=https://docs.docker.com
After=network-online.target firewalld.service
Wants=network-online.target

[Service]
Type=notify
# the default is not to use systemd for cgroups because the delegate issues still
# exists and systemd currently does not support the cgroup feature set required
# for containers run by docker
ExecStart=/usr/bin/dockerd --log-opt max-size=50m --log-opt max-file=10 --storage-driver=devicemapper
ExecReload=/bin/kill -s HUP $MAINPID
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=infinity
LimitNPROC=infinity
LimitCORE=infinity
# Uncomment TasksMax if your systemd version supports it.
# Only systemd 226 and above support this version.
#TasksMax=infinity
TimeoutStartSec=0
# set delegate yes so that systemd does not reset the cgroups of docker containers
Delegate=yes
# kill only the docker process, not all processes in the cgroup
KillMode=process
# restart the docker process if it exits prematurely
Restart=on-failure
RestartSec=10
StartLimitBurst=20
StartLimitInterval=300

[Install]
WantedBy=multi-user.target
```

The changes that I did was to add `RestartSec=10`, modify `StartLimitBurst=20` and `StartLimitInterval=300`.  The docker service will make 20 attempts at an interval of 10 seconds and this whole thing must be under 300 seconds. I was able to verify that Docker service `journalctl -xe -u docker` failed on 1st attempt and then it waited for 10 seconds and then OS was able to start LVM and then docker started. I just kept 20 iterations within 10 seconds interval and that is plenty time for LVM to be available for docker. 

Thanks.
